src/os/bluestore/Btree2Allocator.h

```diff
@@ -0,0 +1,425 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <mutex>
+
+#include "include/cpp-btree/btree_map.h"
+#include "include/cpp-btree/btree_set.h"
+
+#include "Allocator.h"
+
+#include "os/bluestore/bluestore_types.h"
+#include "include/mempool.h"
+#include "common/ceph_mutex.h"
+
+template <size_t BASE_BITS, size_t NUM_BUCKETS, size_t MUX = 1>
+struct SizeBucketCollectionBase {
+  static const size_t base_bits = BASE_BITS;
+  static const size_t base = 1ull << base_bits;
+  static const size_t mux = MUX;
+  static const size_t num_size_buckets = NUM_BUCKETS;
+  static const size_t bucket_max = base << ((mux * (num_size_buckets - 2)));
+
+  inline size_t _get_size_bucket(uint64_t len) const {
+    size_t idx;
+    if (len <= base) {
+      idx = 0;
+    } else if (len > bucket_max) {
+      idx = num_size_buckets - 1;
+    } else {
+      size_t most_bit = cbits(uint64_t(len - 1)) - 1;
+      idx = 1 + ((most_bit - base_bits) / mux);
+    }
+    ceph_assert(idx < num_size_buckets);
+    return idx;
+  }
+};
+
+class ChunkCache {
+public:
+  class LocklessBuf {
+    std::atomic<size_t> ref = 0;
+    size_t count = 0;
+    std::vector<uint32_t> data;
+  public:
+    void init(size_t size) {
+      ceph_assert(data.size() == 0);
+      data.resize(size);
+    }
+    bool maybe_put(uint32_t& v) {
+      bool done = ++ref == 1 && count < data.size();
+      if (done) {
+        data[count++] = v;
+      }
+      --ref;
+      return done;
+    }
+    bool maybe_get(uint32_t& v) {
+      bool done = ++ref == 1 && count > 0;
+      if (done) {
+        v = data[--count];
+      }
+      --ref;
+      return done;
+    }
+    void foreach(std::function<void(uint32_t)> notify) {
+      for (size_t i = 0; i < count; i++) {
+        notify(data[i]);
+      }
+    }
+  };
+
+  static const size_t base_bits = 12; // 4K
+  static const uint64_t base_mask = (1ull << base_bits) - 1;
+  static const size_t num_buckets = 16; // [4K, 8K, 12K, ... 64K] - 16 buckets total
+  static const size_t num_chunks = 16;
+  ChunkCache() {
+    for (size_t i = 0; i < buckets.size(); i++) {
+      buckets[i].init(num_chunks);
+    }
+  }
+  bool maybe_put(uint64_t offset, uint64_t len) {
+    if (!lock.try_lock_shared()) {
+      return false;
+    }
+    bool ret = false;
+    ceph_assert(!(offset & base_mask));
+    ceph_assert(!(len & base_mask));
+    // permit caching chunks with offsets fitting (without LSBs)
+    // into 32 - bit value
+    offset = offset >> base_bits;
+    size_t idx = len >> base_bits;
+    if (offset <= std::numeric_limits<uint32_t>::max() &&
```

Comment:
Some explanantion what the role of `ChunkCache` would be nice.
Without it, I cannot wrap my head around why we are so ok to drop regions with `offset <= std::numeric_limits<uint32_t>::max() << base_bits`